copy/copy.go

```diff
@@ -390,6 +410,33 @@ func compareImageDestinationManifestEqual(ctx context.Context, options *Options,
 	return true, destManifest, destManifestType, destManifestDigest, nil
 }
 
+// removeInstanceFromList removes the given image from the list
+// this should probably move to an internal API
+func removeInstancesFromList(manifestList manifest.List, imageIndices map[int]bool) error {
```

Comment:
No; let’s build the infrastructure the right way rather than piling on tech debt.

It will never be any easier to get that done, and each such “small temporary thing” makes it harder.

Reply:
I appreciate your way of thinking, but its just that I'm a bit lost at how to do it. I can't move the public interface of course to not break it. I can extend it (embed it in an internal one). But I also can't move the implementation structs I guess because those also seem to be public. I can extend those with the internal interface, but then the methods would still be public on the struct and I'm not sure if that's ok. And is it possible in go to do something like return type overloading to get methods like Clone() right?
With some pointers I might be able to perform the refactoring but at this point I'm a bit stuck.

Reply:
Yes, this is the hard part of all this work.

Unlike interfaces, adding new methods to structs is ABI-safe. So my first guess is that the `manifest.List` interface, and the methods that create that (`ListFromBlob`) should be ~duplicated in internal packages. The implementations would _probably_ also be moved to internal-only, with public aliases (`type OCI1Index = internal.OCI1Index`), but I’m _much_ less certain about this part.

---

Alternatively… I think we’ll need to make these refactors for other purposes (adding/removing zstd-compressed variants) fairly soon anyway (but I’m not exactly sure when). So it might be an option to wait with this PR until that infrastructure work is done by other people.

Cc: @vrothberg .

Reply:
I tried but didn't succeed. Apparently go requires return types of a method to be equal in order to implement an interface, a subtype is not supported (I think its the lack of covariance support as described here: https://github.com/golang/go/issues/30602). So the `Clone` method for example has to return a public `List` and not an internal one. But it also doesn't allow a more specific `ListUpdate` to be returned for example.
One option I considered was to add methods like `CloneInternal()` to the private interface. But that seems to be impossible without introducing package cycles because the public interface and public creator method are in the same package. The only workaround/hack I can think of is to do some dynamic registration to redirect the public creator to the private creator.
On the other hand, while I learned more about go with this exercise there's still a lot more for me to learn so I might have overlooked some other options. So unless someone can give me a clear pointer to something I missed I think I'll give up for now.